File: solutions/elixir/017-remove-nth-node-from-end-of-list.ex
# Problem 17: Remove Nth Node From End of List (LeetCode 19)
# Difficulty: Med
# Language: Elixir
# 
defmodule RemoveNth do
  def remove_nth_from_end(list, n) do
    len = length(list)
    idx = len - n
    list
    |> Enum.with_index()
    |> Enum.reject(fn {_, i} -> i == idx end)
    |> Enum.map(&elem(&1, 0))
  end
end

File: solutions/rust/017-remove-nth-node-from-end-of-list.rs
/**
 * Problem 17: Remove Nth Node From End of List (LeetCode 19)
 * Difficulty: Med
 * Language: Rust
 */
pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = &head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = &node.next;
    }
    let idx = vals.len() - n as usize;
    vals.remove(idx);
    vals.into_iter().rev().fold(None, |next, val| {
        Some(Box::new(ListNode { val, next }))
    })
}
